larpandoracontent/LArControlFlow/StitchingCosmicRayMergingTool.cc

```diff
@@ -532,206 +533,115 @@ void StitchingCosmicRayMergingTool::StitchPfos(const MasterAlgorithm *const pAlg
+            {
+                if (iterI == iterJ)
+                    continue;
 
-            const unsigned int twoDHitsPfo1(LArPfoHelper::GetNumberOfTwoDHits(pPfoToShift1)), twoDHitsPfo2(LArPfoHelper::GetNumberOfTwoDHits(pPfoToShift2));
+                const ParticleFlowObject *const pPfoJ(*iterJ);
+                const LArTPC *const pLArTPCJ(pfoToLArTPCMap.at(pPfoJ));
 
-            PfoToLArTPCMap::const_iterator iter1(pfoToLArTPCMap.find(pPfoToShift1));
-            PfoToLArTPCMap::const_iterator iter2(pfoToLArTPCMap.find(pPfoToShift2));
+                if (!LArStitchingHelper::CanTPCsBeStitched(*pLArTPCI, *pLArTPCJ))
+                    continue;
 
-            if ((iter1 != pfoToLArTPCMap.end()) && (iter2 != pfoToLArTPCMap.end()))
-            {
-                const LArTPC *const pLArTPC1(iter1->second);
-                const LArTPC *const pLArTPC2(iter2->second);
+                PfoList pfosToShift({pPfoI, pPfoJ});
 
-                if (((twoDHitsPfo1 + twoDHitsPfo2) > totalHitsStitched) && LArStitchingHelper::CanTPCsBeStitched(*pLArTPC1, *pLArTPC2))
+                for (const ParticleFlowObject *const pPfoToShift : pfosToShift)
                 {
-                    totalHitsStitched = twoDHitsPfo1 + twoDHitsPfo2;
-                    reducedPfoVector.clear();
-                    reducedPfoVector.push_back(pPfoToShift1);
-                    reducedPfoVector.push_back(pPfoToShift2);
+                    // if pfo has already been shifted
+                    if (std::find(shiftedPfos.begin(), shiftedPfos.end(), pPfoToShift) != shiftedPfos.end())
+                        continue;
+
+                    if (!m_useXcoordinate || m_alwaysApplyT0Calculation)
+                    {
+                        // get pfo stitching vertex
+                        const PfoToPointingVertexMatrix::iterator pfoToPointingVertexMatrixIter(pfoToPointingVertexMatrix.find(pPfoToShift));
+                        LArPointingCluster::Vertex stitchingVertex;
+                        
+                        for (const ParticleFlowObject *const pPfo : pfosToShift)
+                        {
+                            if (pPfo == pPfoToShift)
+                                continue;
+
+                            stitchingVertex = pfoToPointingVertexMatrixIter->second.at(pPfo);
+                        }
+
+                        // Determine shift sign from the relative position of stitching vertex and closest TPC boundary position
+                        const LArTPC *const pShiftLArTPC(pfoToLArTPCMap.at(pPfoToShift));
+                        const float tpcBoundaryCenterX(LArStitchingHelper::GetTPCBoundaryCenterX(*pLArTPCI, *pLArTPCJ));
+                        float tpcBoundaryX(0.f);
+
+                        if (pShiftLArTPC->GetCenterX() < tpcBoundaryCenterX)
+                        {
+                            tpcBoundaryX = pShiftLArTPC->GetCenterX() + (pShiftLArTPC->GetWidthX()/2);
```

Comment:
Typically `(pShiftLArTPC->GetWidthX() / 2.f)` but purely stylistic - compiler will do exactly the same thing :)